examples/step-81/step-81.cc

```diff
@@ -0,0 +1,893 @@
+    // and
+    // \f{align}{
+    // i\int_\Omega J_a \cdot \bar{\varphi_i}\text{d}x
+    // - \int_\Omega \mu_r^{-1} \cdot (\nabla \times \bar{\varphi_i}) \text{d}x.
+    // \f}
+    // In doing so, we need test functions $\varphi_i$ and $\varphi_j$, and the
+    // curl of these test variables. We must be careful with the signs of the
+    // imaginary parts of these complex test variables. Moreover, we have a
+    // conditional that changes the parameters if the cell is in the PML region.
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        fe_values.reinit(cell);
+        FEValuesViews::Vector<dim> real_part(fe_values, 0);
+        FEValuesViews::Vector<dim> imag_part(fe_values, dim);
+
+        cell_matrix = 0.;
+        cell_rhs    = 0.;
+
+        cell->get_dof_indices(local_dof_indices);
+        const auto id = cell->material_id();
+
+        const auto &quadrature_points = fe_values.get_quadrature_points();
+
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            const Point<dim> &position = quadrature_points[q_point];
+            const auto        radius   = position.norm();
+            const auto inner_radius    = perfectly_matched_layer.inner_radius;
+
+            auto       mu_inv  = parameters.mu_inv(position, id);
+            auto       epsilon = parameters.epsilon(position, id);
+            const auto J_a     = parameters.J_a(position, id);
+
+            if (radius >= inner_radius)
+              {
+                auto A = perfectly_matched_layer.a_matrix(position);
+                auto B = perfectly_matched_layer.b_matrix(position);
+                auto d = perfectly_matched_layer.d(position);
+
+                mu_inv  = mu_inv / d;
+                epsilon = invert(A) * epsilon * invert(B);
+              };
+
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              {
+                const auto phi_i = real_part.value(i, q_point) -
+                                   1.0i * imag_part.value(i, q_point);
+                const auto curl_phi_i = real_part.curl(i, q_point) -
+                                        1.0i * imag_part.curl(i, q_point);
+
+                const auto rhs_value =
+                  (1.0i * scalar_product(J_a, phi_i)) * fe_values.JxW(q_point);
+                cell_rhs(i) += rhs_value.real();
+
+                for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                  {
+                    const auto phi_j = real_part.value(j, q_point) +
+                                       1.0i * imag_part.value(j, q_point);
+                    const auto curl_phi_j = real_part.curl(j, q_point) +
+                                            1.0i * imag_part.curl(j, q_point);
+
+                    const auto temp =
+                      (scalar_product(mu_inv * curl_phi_j, curl_phi_i) -
+                       scalar_product(epsilon * phi_j, phi_i)) *
+                      fe_values.JxW(q_point);
```

Comment:
This is not something that we will fix in the short run, and nothing that blocks this tutorial program. However, we need to make progress on #9545 for Nedelec as well, because anything but the lowest order Nedelec element this code will be just terribly inefficient. We are talking about a factor of probably 50-100 away from what is possible, when I consider some measurements we made with #13591 for Raviart-Thomas.